src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -474,6 +681,150 @@ protected virtual void ProcessObjectValue()
+        /// once it returns the reader will be returned to the position before the method was called.
+        /// The reader is always positioned on a start object when this method is called.
+        /// </remarks>
+        protected virtual async Task ProcessObjectValueAsync()
+        {
+            Debug.Assert(this.currentBufferedNode.NodeType == JsonNodeType.StartObject, "this.currentBufferedNode.NodeType == JsonNodeType.StartObject");
+            Debug.Assert(this.parsingInStreamError, "this.parsingInStreamError");
+            this.AssertBuffering();
+            this.AssertAsynchronous();
+
+            // Only check for in-stream errors if the buffering reader is told to do so
+            if (!this.DisableInStreamErrorDetection)
+            {
+                // Move to the first property of the potential error object (or the end-object node if no properties exist)
+                await this.ReadInternalAsync()
+                    .ConfigureAwait(false);
+
+                // We only consider this to be an in-stream error if the object has a single 'error' property
+                bool errorPropertyFound = false;
+                Tuple<bool, ODataError> readInStreamErrorPropertyResult = null;
+
+                while (this.currentBufferedNode.NodeType == JsonNodeType.Property)
+                {
+                    // NOTE: the JSON reader already ensures that the value of a property node (which is the name of the property) is a string
+                    string propertyName = (string)this.currentBufferedNode.Value;
+
+                    // If we found any other property than the expected in-stream error property, we don't treat it as an in-stream error
+                    if (string.CompareOrdinal(this.inStreamErrorPropertyName, propertyName) != 0 || errorPropertyFound)
+                    {
+                        return;
+                    }
+
+                    errorPropertyFound = true;
+
+                    // Position the reader over the property value
+                    await this.ReadInternalAsync()
+                        .ConfigureAwait(false);
+
+                    readInStreamErrorPropertyResult = await this.TryReadInStreamErrorPropertyValueAsync()
+                        .ConfigureAwait(false);
+
+                    if (!readInStreamErrorPropertyResult.Item1)
+                    {
+                        // This means we thought we saw an in-stream error, but then
+                        // we didn't see an intelligible error object, so we give up an reading the in-stream error
+                        // and return. We will fail later in some other way. This payload is totally messed up.
+                        return;
+                    }
+                }
+
+                if (errorPropertyFound)
+                {
+                    throw new ODataErrorException(readInStreamErrorPropertyResult?.Item2);

```

Comment:
```suggestion
                    throw new ODataErrorException(readInStreamErrorPropertyResult?.error);

```

Reply:
Can `readInStreamErrorPropertyResult` be null at this point? And if so, what would be the outcome of passing `null` to `ODataErrorException`?

Reply:
@habbes No. It can't be null so I'll modify the logic to take that into account. However, passing a `null` value as the argument doesn't cause a problem